Dj/incremental

IsaacG18 opened the pull request:
Update Cheri pointers to boundary to current header while iterating through in mark process

IsaacG18 committed:
Update outbounds CHERI capability for dirty pages

IsaacG18 committed:
Add Macro to move to the next blk

IsaacG18 committed:
Update to only have a macro to update the bounds

IsaacG18 committed:
fix issue with missing semi-colons

IsaacG18 committed:
get the correct header of the block

IsaacG18 committed:
Update style and remove used macro

djichthys commented on include/private/gc_priv.h:
```diff
@@ -2270,6 +2270,23 @@ ptr_t GC_save_regs_in_stack(void);
+      && (cheri_perms_get(cap) & (CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP))   \
+              != 0)
+# define INBOUND_CAPABILITY(h)                                         \
+    do{                                                                \
```
space between `do` and `{`

djichthys commented on include/private/gc_priv.h:
```diff
@@ -2270,6 +2270,23 @@ ptr_t GC_save_regs_in_stack(void);
+# define INBOUND_CAPABILITY(h)                                         \
+    do{                                                                \
+      word base_addr = cheri_base_get(h);                              \
+      if (!VALID_CAPABILITY(h, base_addr)){                            \
```
space between  closing `)` and `{`  
Also,  a *user* of  the macro may pass in an expression rather than a simple *L-value*. Might be  a good idea to enclose references to `h`  in brackets , i.e.  `(h)`

djichthys commented on include/private/gc_priv.h:
```diff
@@ -2270,6 +2270,23 @@ ptr_t GC_save_regs_in_stack(void);
+      && ADDR(cap) < base_addr + cheri_length_get(cap)                      \
+      && (cheri_perms_get(cap) & (CHERI_PERM_LOAD | CHERI_PERM_LOAD_CAP))   \
+              != 0)
+# define INBOUND_CAPABILITY(h)                                         \
```
I think this should be renamed from `INBOUND_CAPABILITY` to something which expresses the purpose of the macro better.  As far as I can see, there are two things going on 
1. check if  calculated capability address is valid and in-bounds. 
2. If yes return the newly calculated bounds value. If NOT, then : 
     a. attempt to obtain a valid capability with the *originally* out-of-bounds capbility from the correct header. If this succeeds, h will be set correctly. 
     b. If 2.a FAILS, then return the invalid capability (We are being optimistic here, but so far, every other location that will utilize the return from this function seems to be making a call to ascertain the *HDR* based on the return value 
     
I think it should be renamed to something like `RESOLVE_CAP()`

djichthys commented on mark.c:
```diff
@@ -2186,9 +2195,18 @@ GC_push_next_marked_dirty(struct hblk *h)
     GC_push_marked(h, hhdr);
   }
+# ifdef CHERI_PURECAP
+  h += OBJ_SZ_TO_BLOCKS(hhdr->hb_sz);
```
Lift this line outside the conditional macro and also the return statements from both parts of the macro. Makes for cleaner code i.e. 
```
   h += OBJ_SZ_TO_BLOCKS(hhdr->hb_sz);
# ifdef CHERI_PURECAP
   RESOLVE_CAP(h); 
# endif 
   return h; 
```

djichthys commented on mark.c:
```diff
@@ -2222,7 +2240,16 @@ GC_push_next_marked_uncollectable(struct hblk *h)
     hhdr = HDR(h);
   }
+# ifdef CHERI_PURECAP
+  h += OBJ_SZ_TO_BLOCKS(hhdr->hb_sz);
```
Same comment as on [line 2204](https://github.com/djichthys/bdwgc/pull/1/files#r2185471357)

IsaacG18 committed:
update the macros name and update the style

djichthys commented on os_dep.c:
```diff
@@ -3479,8 +3479,9 @@ GC_write_fault_handler(struct _EXCEPTION_POINTERS *exc_info)
 #    ifdef CHERI_PURECAP
-    hdr *hhdr = HDR(h); 
-    UNPROTECT(hhdr->hb_block, GC_page_size);
+    hdr *hhdr = HDR(GC_prev_block(h));
```
I think this change is actually masking the issue -- from my debugging the signal handler is actually not returning the address that triggered the fault handler in the first place.

Commit:
Resolve capablities in signal handler using fallback

On CHERIBSD, fallback to using global heap capbilities
when a valid header cannot be obtained for the address
triggering the signal.
